ql/src/java/org/apache/hadoop/hive/ql/stats/StatsUtils.java

```diff
@@ -1736,6 +1749,13 @@ private static boolean isWideningCast(ExprNodeGenericFuncDesc engfd) {
             engfd.getTypeInfo());
   }
 
+  private static boolean isColStatsReusableFunc(ExprNodeGenericFuncDesc engfd) {
+    GenericUDF udf = engfd.getGenericUDF();
+    Class udfClass =
+        (udf instanceof GenericUDFBridge) ? ((GenericUDFBridge)udf).getUdfClass() : udf.getClass();
+    return udfClass == GenericUDFLower.class || udfClass == GenericUDFUpper.class;
```

Comment:
I like this PR. I have also observed that an estimation was wrong, and Map Join was disabled.
We implement some estimation APIs on the UDF side. I'll share examples with you. This is not a strong suggestion, and the current implementation seems to be a good starting point. We may finally need a pluggable I/F to empower custom UDFs more.
- https://github.com/apache/hive/blob/rel/release-4.0.0/ql/src/java/org/apache/hadoop/hive/ql/udf/generic/NDV.java
- https://github.com/apache/hive/blob/rel/release-4.0.0/ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDAFEvaluator.java#L63-L81

Reply:
Thanks for sharing the examples. I moved the estimation logic to UDF side by using `StatEstimatorProvider` interface.

Reply:
I didn't know the interface and looks super nice!